### PR TITLE
live_migration: Add case about execute libvirt statistics command

### PR DESCRIPTION
**Pull request by cliping**
RHEL7-17314 - [Migration] Execute libvirt statistics command
during live migration - domblkinfo/domblklist/dommemstat

Signed-off-by: lcheng <lcheng@redhat.com>

### CODE REVIEW

**Comment by cliping** [comment]
Test result:

#avocado run --vt-type libvirt --vt-machine-type q35 virsh.live_migration.execute_statistics_cmd.p2p_live.without_postcopy --vt-connect-uri qemu:///system
JOB ID     : 63da2e6c62174ee3fac4aa828976341b54334f11
JOB LOG    : /root/avocado/job-results/job-2022-03-15T06.57-63da2e6/job.log
 (1/1) type_specific.io-github-autotest-libvirt.virsh.live_migration.execute_statistics_cmd.p2p_live.without_postcopy: PASS (166.79 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 167.50 s

**Comment by cliping** [comment]
@Yingshun @fangge1212 , please help to review, thanks.

File: provider/migration/migration_base.py [hunk 1]
```diff
@@ -227,3 +228,25 @@ def set_migrate_speed_to_high(params):
...
+
+    vmxml = vm_xml.VMXML.new_from_dumpxml(vm_name)
+    disks = vmxml.get_disk_all_by_expr('type==block', 'device==disk')
+    logging.info("disks: %s" % disks)
+    for disk in list(disks.values()):
+        disk_source = disk.find('source').get('dev')
+        disk_target = disk.find('target').get('dev')
+        logging.info("disk_source: %s" % disk_source)
+        logging.info("disk_target: %s" % disk_target)
+    virsh.domblkstat(vm_name, disk_target, "", ignore_status=False, debug=True)
```
Comment (chunfuwen) [inline provider/migration/migration_base.py:249]: Does line 249 to 252 need put under for loop umbrella on line 244?

File: provider/migration/migration_base.py [hunk 2]
```diff
@@ -227,3 +228,25 @@ def set_migrate_speed_to_high(params):
...
+    virsh.domblkinfo(vm_name, disk_source, ignore_status=False, debug=True)
```
Comment (chunfuwen) [inline provider/migration/migration_base.py:250]: debug_kargs = {'ignore_status': False,  'debug' : True}
virsh.domblkinfo(vm_name, disk_source, **debug_kargs)
this way help simplify repeated parameters afterwards

File: libvirt/tests/src/migration/live_migration.py [hunk 3]
```diff
@@ -117,6 +119,37 @@ def recover_config_file(conf_obj, params):
...
+def prepare_block_device(params, vm):
+    """
+    Prepare block_device for test
+
+    :param params: dict, parameters used
+    :param vm: vm object
+    """
+    block_device = params.get("nfs_mount_dir") + '/' + params.get("block_device_name")
+    image_cmd = "qemu-img create -f qcow2 %s 1G" % block_device
+    process.run(image_cmd, shell=True)
```
Comment (chunfuwen) [inline libvirt/tests/src/migration/live_migration.py:131]: we have libvirt.create_local_image() to help achieve this

File: libvirt/tests/src/migration/live_migration.py [hunk 4]
```diff
@@ -117,6 +119,37 @@ def recover_config_file(conf_obj, params):
...
+    # Find a free loop device on target
+    cmd_result = remote_old.run_remote_cmd("losetup --find", params, ignore_status=False)
+    target_loop_dev = cmd_result.stdout_text.strip()
+    logging.info("target loop dev: %s" % target_loop_dev)
+    # Setup a loop device on target
+    cmd = 'losetup %s %s' % (target_loop_dev, block_device)
+    remote_old.run_remote_cmd(cmd, params, ignore_status=False)
+
+    # Attach block device to guest
+    vm.attach_disk(source_loop_dev, debug=True)
```
Comment (chunfuwen) [inline libvirt/tests/src/migration/live_migration.py:149]: In current tp-libvirt code, mostly we use virsh.attach_disk()

File: provider/migration/migration_base.py (hunk 1 above)
Comment (cliping) [reply provider/migration/migration_base.py]: Updated, thanks.

File: provider/migration/migration_base.py (hunk 2 above)
Comment (cliping) [reply provider/migration/migration_base.py]: Updated, thanks.

File: libvirt/tests/src/migration/live_migration.py (hunk 3 above)
Comment (cliping) [reply libvirt/tests/src/migration/live_migration.py]: Updated to libvirt.create_local_disk(), thanks.

File: libvirt/tests/src/migration/live_migration.py (hunk 4 above)
Comment (cliping) [reply libvirt/tests/src/migration/live_migration.py]: Updated to virsh.attach_disk(), thanks.

**Review by chunfuwen — APPROVED** [review]
LGTM

File: provider/migration/migration_base.py [hunk 5]
```diff
@@ -227,3 +228,25 @@ def set_migrate_speed_to_high(params):
...
+
+    vmxml = vm_xml.VMXML.new_from_dumpxml(vm_name)
+    disks = vmxml.get_disk_all_by_expr('type==block', 'device==disk')
+    logging.info("disks: %s" % disks)
+    debug_kargs = {'ignore_status': False, 'debug': True}
+    for disk in list(disks.values()):
+        disk_source = disk.find('source').get('dev')
+        disk_target = disk.find('target').get('dev')
+        logging.info("disk_source: %s" % disk_source)
+        logging.info("disk_target: %s" % disk_target)
```
Comment (Yingshun) [inline provider/migration/migration_base.py:248]: Same as above, I think they're be debug level messages.

File: libvirt/tests/src/migration/live_migration.py [hunk 6]
```diff
@@ -117,6 +119,36 @@ def recover_config_file(conf_obj, params):
...
+    # Find a free loop device on target
+    cmd_result = remote_old.run_remote_cmd("losetup --find", params, ignore_status=False)
+    target_loop_dev = cmd_result.stdout_text.strip()
+    logging.info("target loop dev: %s" % target_loop_dev)
+    # Setup a loop device on target
+    cmd = 'losetup %s %s' % (target_loop_dev, block_device)
+    remote_old.run_remote_cmd(cmd, params, ignore_status=False)
+
+    # Attach block device to guest
+    virsh.attach_disk(vm_name, source_loop_dev, 'vdb', debug=True)
```
Comment (Yingshun) [inline libvirt/tests/src/migration/live_migration.py:148]: Result checking is missing I think. Again, I think we should not hard-code 'vdb' here, could you update it? Thanks!

Comment (cliping) [reply libvirt/tests/src/migration/live_migration.py]: Updated.

File: provider/migration/migration_base.py (hunk 5 above)
Comment (cliping) [reply provider/migration/migration_base.py]: Updated.

File: provider/migration/migration_base.py [hunk 7]
```diff
@@ -227,3 +228,25 @@ def set_migrate_speed_to_high(params):
...
+def execute_statistics_command(vm_name):
+    """
+    Execute statistics command
+
+    :param vm_name: guest name
+    """
+
+    vmxml = vm_xml.VMXML.new_from_dumpxml(vm_name)
+    disks = vmxml.get_disk_all_by_expr('type==block', 'device==disk')
+    logging.info("disks: %s" % disks)
```
[the remark this answers is not in the document]
Comment (cliping) [reply provider/migration/migration_base.py]: Updated.

File: libvirt/tests/src/migration/live_migration.py [hunk 8]
```diff
@@ -117,6 +119,36 @@ def recover_config_file(conf_obj, params):
...
+
+    :param params: dict, parameters used
+    :param vm_name: vm name
+    """
+    block_device = params.get("nfs_mount_dir") + '/' + params.get("block_device_name")
+    libvirt.create_local_disk("file", block_device, '1', "qcow2")
+
+    # Find a free loop device on source
+    source_loop_dev = process.run("losetup --find", shell=True).stdout_text.strip()
+    logging.info("source loop dev: %s" % source_loop_dev)
```
[the remark this answers is not in the document]
Comment (cliping) [reply libvirt/tests/src/migration/live_migration.py]: Updated.

File: libvirt/tests/src/migration/live_migration.py [hunk 9]
```diff
@@ -117,6 +119,36 @@ def recover_config_file(conf_obj, params):
...
+    cmd = 'losetup %s %s' % (source_loop_dev, block_device)
+    process.run(cmd, shell=True)
+
+    # Find a free loop device on target
+    cmd_result = remote_old.run_remote_cmd("losetup --find", params, ignore_status=False)
+    target_loop_dev = cmd_result.stdout_text.strip()
+    logging.info("target loop dev: %s" % target_loop_dev)
+    # Setup a loop device on target
+    cmd = 'losetup %s %s' % (target_loop_dev, block_device)
+    remote_old.run_remote_cmd(cmd, params, ignore_status=False)
```
[the remark this answers is not in the document]
Comment (cliping) [reply libvirt/tests/src/migration/live_migration.py]: Updated, thanks.

File: provider/migration/migration_base.py [hunk 10]
```diff
@@ -227,3 +228,25 @@ def set_migrate_speed_to_high(params):
...
+
+def execute_statistics_command(vm_name):
+    """
+    Execute statistics command
+
+    :param vm_name: guest name
+    """
+
+    vmxml = vm_xml.VMXML.new_from_dumpxml(vm_name)
+    disks = vmxml.get_disk_all_by_expr('type==block', 'device==disk')
```
[the remark this answers is not in the document]
Comment (cliping) [reply provider/migration/migration_base.py]: Updated.

**Comment by cliping** [comment]
@Yingshun , updated code, please help to review. Thanks.

File: libvirt/tests/src/migration/live_migration.py [hunk 11]
```diff
@@ -117,6 +120,53 @@ def recover_config_file(conf_obj, params):
                                      config_object=conf_obj)
 
 
+def setup_loop_dev(block_device, remote=False, remote_params=None):
```
Comment (Yingshun) [inline libvirt/tests/src/migration/live_migration.py:123]: I wonder if we can delete 'remote' and determine if it's a remote operations according to 'remote_params'?

File: libvirt/tests/src/migration/live_migration.py [hunk 12]
```diff
@@ -117,6 +120,53 @@ def recover_config_file(conf_obj, params):
...
+                          server_ip, server_user, server_pwd
+    """
+    loop_dev = None
+    # Find a free loop device
+    cmd = "losetup --find"
+    if remote:
+        loop_dev = remote_old.run_remote_cmd(cmd, remote_params, ignore_status=False).stdout_text.strip()
+    else:
+        loop_dev = process.run(cmd, shell=True).stdout_text.strip()
+    logging.debug("loop dev: %s" % loop_dev)
```
Comment (Yingshun) [inline libvirt/tests/src/migration/live_migration.py:139]:
```suggestion
    logging.debug("loop dev: %s", loop_dev)
```

File: libvirt/tests/src/migration/live_migration.py [hunk 13]
```diff
@@ -117,6 +120,53 @@ def recover_config_file(conf_obj, params):
...
+    block_device = params.get("nfs_mount_dir") + '/' + params.get("block_device_name")
+    libvirt.create_local_disk("file", block_device, '1', "qcow2")
+
+    source_loop_dev = setup_loop_dev(block_device)
+    target_loop_dev = setup_loop_dev(block_device, remote=True, remote_params=params)
+
+    # Attach block device to guest
+    result = virsh.attach_disk(vm_name, source_loop_dev, target_dev, debug=True)
+    utils_misc.wait_for(lambda: (result.stdout ==
+                        "Disk attached successfully"), 10)
```
Comment (Yingshun) [inline libvirt/tests/src/migration/live_migration.py:166]: The value of 'result' is obtained already, I don't quite understand why it takes 10s to compare stdout's output? Or should it be something like utils_misc.wait_for(lambda: (virsh.attach_disk(vm_name, source_loop_dev, target_dev, debug=True).stdout_text == "Disk attached successfully'), 10)?

File: provider/migration/migration_base.py [hunk 14]
```diff
@@ -227,3 +228,27 @@ def set_migrate_speed_to_high(params):
...
+    Execute statistics command
+
+    :param params: dict, used to setup the connection
+    """
+    vm_name = params.get("migrate_main_vm")
+    disk_type = params.get("disk_type")
+
+    vmxml = vm_xml.VMXML.new_from_dumpxml(vm_name)
+    disks = vmxml.get_disk_all_by_expr('type==%s' % disk_type, 'device==disk')
+    logging.debug("disks: %s" % disks)
```
Comment (Yingshun) [inline provider/migration/migration_base.py:244]:
```suggestion
    logging.debug("disks: %s", disks)
```

File: provider/migration/migration_base.py [hunk 15]
```diff
@@ -227,3 +228,27 @@ def set_migrate_speed_to_high(params):
...
+    debug_kargs = {'ignore_status': False, 'debug': True}
+    for disk in list(disks.values()):
+        disk_source = disk.find('source').get('dev')
+        disk_target = disk.find('target').get('dev')
+        logging.debug("disk_source: %s" % disk_source)
+        logging.debug("disk_target: %s" % disk_target)
```
Comment (Yingshun) [inline provider/migration/migration_base.py:250]:
```suggestion
        logging.debug("disk_source: %s", disk_source)
        logging.debug("disk_target: %s", disk_target)
```

File: libvirt/tests/src/migration/live_migration.py (hunk 13 above)
Comment (cliping) [reply libvirt/tests/src/migration/live_migration.py]: Updated to utils_misc.wait_for(lambda: not virsh.attach_disk(vm_name, source_loop_dev, target_dev, debug=True).exit_status, 10), thanks.

File: libvirt/tests/src/migration/live_migration.py (hunk 11 above)
Comment (cliping) [reply libvirt/tests/src/migration/live_migration.py]: Updated, thanks.

File: provider/migration/migration_base.py (hunk 14 above)
Comment (cliping) [reply provider/migration/migration_base.py]: Updated.

File: provider/migration/migration_base.py (hunk 15 above)
Comment (cliping) [reply provider/migration/migration_base.py]: Updated.

File: libvirt/tests/src/migration/live_migration.py (hunk 12 above)
Comment (cliping) [reply libvirt/tests/src/migration/live_migration.py]: Updated.

**Review by Yingshun — APPROVED** [review]
LGTM

**Review by fangge1212 — APPROVED** [review]

**Pull request merged**